numba/targets/arraymath.py

```diff
@@ -668,6 +668,83 @@ def _trivial_isnan(x):
         return _trivial_isnan
 
 
+@overload(np.iscomplex)
+def np_iscomplex(x):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    dt = determine_dtype(x)
+    if np.issubdtype(dt, np.complexfloating):
+        def impl(x):
+            ax = np.asarray(x) # NumPy uses asanyarray here!
+            return ax.imag != 0
+    else:
+        def impl(x):
+            ax = np.asarray(x)
+            zeros = np.zeros(ax.shape, dtype=types.boolean)
+            return zeros[()] # hack when x is a scalar value
+    return impl
+
+
+@overload(np.isreal)
+def np_isreal(x):
+    def impl(x):
+        ax = np.iscomplex(x)
+        return np.invert(ax)
+    return impl
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    dt = determine_dtype(x)
+    if np.issubdtype(dt, np.complexfloating):
+        @register_jitable
+        def impl(x):
+            return True
+    else:
+        @register_jitable
+        def impl(x):
+            return False
+    return impl
```

Comment:
Pretty sure this can just be:
```python
@overload(np.iscomplexobj)
def iscomplexobj(x):
    dt = determine_dtype(x)
    iscmplx = np.issubdtype(dt, np.complexfloating)
    return lambda _: iscmplx
```
I think the return value of `iscomplexobj` can be determined solely by type, and therefore resolved entirely at typing time, consequently the implementation can just close over the result. Also, specialisations don't need to be `@register_jitable`, `@overload` expects a standard Python function to be returned and will just wrap it as jit-able.

Reply:
> Also, specializations don't need to be @register_jitable, @overload expects a standard Python function to be returned and will just wrap it as jit-able.
I didn't know that. 

Your solution works but I had to change the last statement into
```python
@overload(np.iscomplexobj)
def iscomplexobj(x):
    dt = determine_dtype(x)
    iscmplx = np.issubdtype(dt, np.complexfloating)
    return lambda x: iscmplx
```